mixer/adapter/cloudwatch/handler.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY Type, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudwatch
+
+import (
+	"errors"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/cloudwatch"
+
+	"istio.io/istio/mixer/template/metric"
+)
+
+const (
+	// CloudWatch enforced limit
+	// https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html
+	metricDatumLimit = 20
+	// CloudWatch enforced limit
+	// https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/cloudwatch_limits.html
+	dimensionLimit = 10
+)
+
+func handleValidMetrics(h *Handler, insts []*metric.Instance) error {
+	metricData := generateMetricData(h, insts)
+	_, err := sendMetricsToCloudWatch(h, metricData)
+	return err
+}
+
+func sendMetricsToCloudWatch(h *Handler, metricData []*cloudwatch.MetricDatum) (int, error) {
+	cloudWatchCallCount := 0
+	var err error
+
+	for i := 0; i < len(metricData); i += metricDatumLimit {
+		size := i + metricDatumLimit
+		if len(metricData) < size {
+			size = len(metricData)
+		}
+
+		err = putMetricData(h, metricData[i:size])
+		if err == nil {
+			cloudWatchCallCount++
+		}
+	}
+
+	return cloudWatchCallCount, err
+}
+
+func generateMetricData(h *Handler, insts []*metric.Instance) []*cloudwatch.MetricDatum {
+	metricData := make([]*cloudwatch.MetricDatum, 0, len(insts))
+
+	for _, inst := range insts {
+		cwMetric := h.cfg.GetMetricInfo()[inst.Name]
+
+		datum := cloudwatch.MetricDatum{
+			MetricName: aws.String(inst.Name),
+			Unit:       aws.String(cwMetric.GetUnit()),
+		}
+
+		if cwMetric.GetStorageResolution() != 0 {
+			datum.SetStorageResolution(cwMetric.GetStorageResolution())
+		}
+
+		value, err := getNumericValue(h, inst)
+		if err != nil {
+			continue
+		}
+		datum.SetValue(value)
+
+		// Only 10 dimenstions can be added to a metric. This is a cloudwatch limit.
+		d := make([]*cloudwatch.Dimension, 0, dimensionLimit)
+
+		for k, v := range inst.Dimensions {
+
+			// check whether the dimension is a timestamp
+			t, ok := v.(time.Time)
+			if ok {
+				datum.SetTimestamp(t)
```

Comment:
are we concerned about (potentially) multiple timestamps (request/response times)?